### ISSUE TITLE
Backpropagation engenders Hebbian learning

### ISSUE THREAD

**Issue by clockwiser** [guillaume-chevalier/Spiking-Neural-Network-SNN-with-PyTorch-where-Backpropagation-engenders-STDP#3]
It is an idea very interesting.

Think about it: from the point of view of a neuron that spikes, upon performing backpropagation, the neuron will backpropagate to its inputs for which the input signal was received more recently than not, hence the signal that really contributed to firing. Why? Because older signals are decayed exponentially and their gradients vanish towards zero. So with SNNs, the gradients are mostly transferred to input that was fired just before self-firing to an output from the point of view of a neuron.

What is the gradient here? 
Generally gradient comes from loss, trying to minimize the loss. But here what is the loss?

**Comment by guillaume-chevalier** [comment]
Cool!

Generally speaking, I think that the loss might perhaps be some kind of reward function related to pain or pleasure. And maybe with some kind of GAN discriminator function of pleasure/pain, but more like in a Reinforcement Learning algorithm. 

Or also I got an idea that it could perhaps make use of brain rhythms. E.g.: upon completing each "rhythmic" cyles, neurons could compute some sort of difference between the last pass and the new pass of information, such that it learns to autoencode information better to reduce the number of cycles needed to process, something akin to Contrastive Divergence (CD) in autoencoders. Like this, the learning is unsupervised and doesn't need a reward function.
